lib/charms/vault_k8s/v0/vault_tls.py

```diff
@@ -0,0 +1,462 @@
+            return content["privatekey"], content["certificate"]
+        except (TypeError, SecretNotFoundError, AttributeError):
+            raise PeerSecretError(secret_name=CA_CERTIFICATE_JUJU_SECRET_KEY)
+
+    def _set_ca_certificate_secret_in_peer_relation(
+        self,
+        private_key: str,
+        certificate: str,
+    ) -> None:
+        """Set the vault CA certificate secret in the peer relation.
+
+        Args:
+            private_key: Private key
+            certificate: certificate
+        """
+        if not self.charm._is_peer_relation_created():
+            raise RuntimeError("Peer relation not created")
+        juju_secret_content = {
+            "privatekey": private_key,
+            "certificate": certificate,
+        }
+        juju_secret = self.charm.app.add_secret(
+            juju_secret_content, label=CA_CERTIFICATE_JUJU_SECRET_LABEL
+        )
+        peer_relation = self.charm.model.get_relation(self.peer_relation)
+        peer_relation.data[self.charm.app].update({CA_CERTIFICATE_JUJU_SECRET_KEY: juju_secret.id})
+        tlslogger.debug("Vault CA certificate secret set in peer relation")
+
+    def _remove_ca_certificate_secret_in_peer_relation(self) -> None:
+        """Removes the vault CA certificate secret in the peer relation.
+
+        Only the key from the relation is removed. The actual juju secret is unmodified.
+        """
+        if not self.charm._is_peer_relation_created():
+            raise RuntimeError("Peer relation not created")
+        peer_relation = self.model.get_relation(self.peer_relation)
+        if peer_relation:
+            peer_relation.data[self.charm.app].pop(CA_CERTIFICATE_JUJU_SECRET_KEY)
+            tlslogger.debug("Vault CA certificate secret removed from peer relation")
+
+    def ca_certificate_set_in_peer_relation(self) -> bool:
+        """Returns whether CA certificate is stored in peer relation data."""
+        try:
+            ca_private_key, ca_certificate = self._get_ca_certificate_secret_in_peer_relation()
+            if ca_private_key and ca_certificate:
+                return True
+        except PeerSecretError:
+            return False
+        return False
+
+    """TODO: The following are methods that go through the container.
+    They will be extracted out in the future."""
+
+    def _remove_all_certs_from_workload(self) -> None:
+        """Removes the certificate files that are used for authentication."""
+        self._remove_tls_file_from_workload(File.CA)
+        self._remove_tls_file_from_workload(File.CERT)
+        self._remove_tls_file_from_workload(File.CSR)
+        tlslogger.debug("Removed existing certificate files from workload.")
+
+    def _reload_vault_container(self) -> None:
+        """Sends a SIGHUP signal to the container which reloads Vault's files. Fails gracefully."""
+        try:
+            self.charm._container.send_signal(SIGHUP, self.charm._container_name)
```

Comment:
How can we assume `charm` to have a `_container` variable?
in `charm.py` in the case of `vault-k8s` it is there because:
`self._container = self.unit.get_container(self._container_name)`, however I'm not sure we can make a general assumption here.

(Sorry if I am commenting on something that has been already discussed.)